// @nadiia/file-loader/dist/index.js
'use strict';

/*
  MIT License http://www.opensource.org/licenses/mit-license.php
  Author Tobias Koppers @sokra
*/
var path = require('path');
var loaderUtils = require('loader-utils');
var storageSingleton = require('./storage-singleton').default;

module.exports = function fileLoader(content) {
  if (!this.emitFile) throw new Error('emitFile is required from module system');

  var options = loaderUtils.getOptions(this) || {};

  var config = {
    publicPath: undefined,
    useRelativePath: false,
    name: '[hash].[ext]',
    storeFile: false,
    storeFileTarget: null
  };

  // options takes precedence over config
  Object.keys(options).forEach(function (attr) {
    config[attr] = options[attr];
  });

  var context = options.context || this.rootContext || this.options && this.options.context;
  var url = loaderUtils.interpolateName(this, config.name, {
    context,
    content,
    regExp: config.regExp
  });

  var outputPath = '';
  if (config.outputPath) {
    // support functions as outputPath to generate them dynamically
    outputPath = typeof config.outputPath === 'function' ? config.outputPath(url) : config.outputPath;
  }

  var filePath = this.resourcePath;
  if (config.useRelativePath) {
    // eslint-disable-next-line no-mixed-operators
    var issuerContext = this._module && this._module.issuer && this._module.issuer.context || context;
    var relativeUrl = issuerContext && path.relative(issuerContext, filePath).split(path.sep).join('/');
    var relativePath = relativeUrl && `${path.dirname(relativeUrl)}/`;
    // eslint-disable-next-line no-bitwise
    if (~relativePath.indexOf('../')) {
      // eslint-disable-line no-bitwise
      outputPath = path.posix.join(outputPath, relativePath, url);
    } else {
      outputPath = relativePath + url;
    }
    url = relativePath + url;
  } else if (config.outputPath) {
    // support functions as outputPath to generate them dynamically
    outputPath = typeof config.outputPath === 'function' ? config.outputPath(url) : config.outputPath + url;
    url = outputPath;
  } else {
    outputPath = url;
  }

  var publicPath = `__webpack_public_path__ + ${JSON.stringify(url)}`;
  if (config.publicPath !== undefined) {
    // support functions as publicPath to generate them dynamically
    publicPath = JSON.stringify(typeof config.publicPath === 'function' ? config.publicPath(url) : config.publicPath + url);
  }

  var storage = storageSingleton.getStorage();
  var storeFile = config.storeFile,
      storeFileTarget = config.storeFileTarget;

  if (options.emitFile === undefined || options.emitFile) {
    // when storeFile param is passed we don't emit a file
    // but store it to be added added later as an additional asset to a compilation
    // it allows adding these files in a different compilation
    if (storeFile && (!storeFileTarget || this.target === storeFileTarget)) {
      storage.addFile(outputPath, content);
    } else {
      this.emitFile(outputPath, content);
    }
  }

  return `module.exports = ${publicPath};`;
};

module.exports.raw = true;
