// fusion-react/dist/browser.es5.es.js
import { createElement, Children, Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import FusionApp, { createPlugin } from 'fusion-core';
import { prepare, middleware } from 'fusion-react-async';

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var clientRender = (function (el) {
  var domElement = document.getElementById('root');

  if (!domElement) {
    throw new Error("Could not find 'root' element");
  }

  return ReactDOM.hydrate ? ReactDOM.hydrate(el, domElement) : ReactDOM.render(el, domElement);
});

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose(subClass, superClass) { subClass.prototype.__proto__ = superClass && superClass.prototype; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var Provider = {
  create: function create(name) {
    var _objectSpread2;

    var Provider =
    /*#__PURE__*/
    function (_React$Component) {
      _inheritsLoose(Provider, _React$Component);

      function Provider() {
        return _React$Component.apply(this, arguments) || this;
      }

      var _proto = Provider.prototype;

      _proto.getChildContext = function getChildContext() {
        var _ref;

        return _ref = {}, _ref[name] = this.props.provides, _ref;
      };

      _proto.render = function render() {
        return Children.only(this.props.children);
      };

      return Provider;
    }(Component);

    Provider.childContextTypes = _objectSpread({}, Provider.childContextTypes || {}, (_objectSpread2 = {}, _objectSpread2[name] = PropTypes.any.isRequired, _objectSpread2));
    Provider.displayName = name.replace(/^./, function (c) {
      return c.toUpperCase();
    }) + 'Provider';
    return Provider;
  }
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var plugin = {
  create: function create(name, plugin, provider) {
    if (plugin.__plugin__ === undefined) {
      plugin = createPlugin(plugin);
    }

    if (!plugin.__plugin__) {
      throw new Error('Provided plugin does not match FusionPlugin<TDeps, TService>');
    }

    var originalMiddleware = plugin.middleware;
    var ProviderComponent = provider || Provider.create(name);

    plugin.middleware = function (deps, provides) {
      var nextMiddleware = originalMiddleware && originalMiddleware(deps, provides);

      var mw = function mw(ctx, next) {
        if (ctx.element) {
          ctx.element = createElement(ProviderComponent, {
            provides: provides,
            ctx: ctx
          }, ctx.element);
        }

        if (nextMiddleware) {
          return nextMiddleware(ctx, next);
        }

        return next();
      };

      return mw;
    };

    return plugin;
  }
};

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$1(target, key, source[key]); }); } return target; }

function _defineProperty$1(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _inheritsLoose$1(subClass, superClass) { subClass.prototype.__proto__ = superClass && superClass.prototype; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var hoc = {
  create: function create(name, mapProvidesToProps) {
    var mapProvides = mapProvidesToProps ? mapProvidesToProps : function (provides) {
      var _ref;

      return _ref = {}, _ref[name] = provides, _ref;
    };
    return function (Component$$1) {
      var _HOC$contextTypes;

      var HOC =
      /*#__PURE__*/
      function (_React$Component) {
        _inheritsLoose$1(HOC, _React$Component);

        function HOC(props, ctx) {
          var _this;

          _this = _React$Component.call(this, props, ctx) || this;
          _this.provides = ctx[name];
          return _this;
        }

        var _proto = HOC.prototype;

        _proto.render = function render() {
          var props = _objectSpread$1({}, this.props, mapProvides(this.provides));

          return createElement(Component$$1, props);
        };

        return HOC;
      }(Component);

      var displayName = Component$$1.displayName || Component$$1.name;
      HOC.displayName = 'With' + name.replace(/^./, function (c) {
        return c.toUpperCase();
      }) + '(' + displayName + ')';
      HOC.contextTypes = (_HOC$contextTypes = {}, _HOC$contextTypes[name] = PropTypes.any.isRequired, _HOC$contextTypes);
      return HOC;
    };
  }
};

function _inheritsLoose$2(subClass, superClass) { subClass.prototype.__proto__ = superClass && superClass.prototype; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var App =
/*#__PURE__*/
function (_FusionApp) {
  _inheritsLoose$2(App, _FusionApp);

  function App(root, render) {
    var renderer = createPlugin({
      provides: function provides() {
        return function (el) {
          return prepare(el).then(function () {
            if (render) {
              return render(el);
            }

            return clientRender(el);
          });
        };
      },
      middleware: function middleware$$1() {
        return middleware;
      }
    });
    return _FusionApp.call(this, root, renderer) || this;
  }

  return App;
}(FusionApp);

export default App;
export { plugin as ProviderPlugin, hoc as ProvidedHOC, Provider };
