// fusion-react/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var server = require('react-dom/server');
var PropTypes = _interopDefault(require('prop-types'));
var FusionApp = require('fusion-core');
var FusionApp__default = _interopDefault(FusionApp);
var fusionReactAsync = require('fusion-react-async');

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */
var serverRender = (el => `<div id='root'>${server.renderToString(el)}</div>`);

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var Provider = {
  create: name => {
    class Provider extends React.Component {
      getChildContext() {
        return {
          [name]: this.props.provides
        };
      }

      render() {
        return React.Children.only(this.props.children);
      }

    }

    Provider.childContextTypes = _objectSpread({}, Provider.childContextTypes || {}, {
      [name]: PropTypes.any.isRequired
    });
    Provider.displayName = name.replace(/^./, c => c.toUpperCase()) + 'Provider';
    return Provider;
  }
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var plugin = {
  create: (name, plugin, provider) => {
    if (plugin.__plugin__ === undefined) {
      plugin = FusionApp.createPlugin(plugin);
    }

    if (!plugin.__plugin__) {
      throw new Error('Provided plugin does not match FusionPlugin<TDeps, TService>');
    }

    let originalMiddleware = plugin.middleware;
    const ProviderComponent = provider || Provider.create(name);

    plugin.middleware = (deps, provides) => {
      let nextMiddleware = originalMiddleware && originalMiddleware(deps, provides);

      const mw = function (ctx, next) {
        if (ctx.element) {
          ctx.element = React.createElement(ProviderComponent, {
            provides,
            ctx
          }, ctx.element);
        }

        if (nextMiddleware) {
          return nextMiddleware(ctx, next);
        }

        return next();
      };

      return mw;
    };

    return plugin;
  }
};

function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty$1(target, key, source[key]); }); } return target; }

function _defineProperty$1(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var hoc = {
  create: (name, mapProvidesToProps) => {
    const mapProvides = mapProvidesToProps ? mapProvidesToProps : provides => ({
      [name]: provides
    });
    return Component => {
      class HOC extends React.Component {
        constructor(props, ctx) {
          super(props, ctx);
          this.provides = ctx[name];
        }

        render() {
          const props = _objectSpread$1({}, this.props, mapProvides(this.provides));

          return React.createElement(Component, props);
        }

      }

      const displayName = Component.displayName || Component.name;
      HOC.displayName = 'With' + name.replace(/^./, c => c.toUpperCase()) + '(' + displayName + ')';
      HOC.contextTypes = {
        [name]: PropTypes.any.isRequired
      };
      return HOC;
    };
  }
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
class App extends FusionApp__default {
  constructor(root, render) {
    const renderer = FusionApp.createPlugin({
      provides() {
        return el => {
          return fusionReactAsync.prepare(el).then(() => {
            if (render) {
              return render(el);
            }

            return serverRender(el);
          });
        };
      },

      middleware() {
        return fusionReactAsync.middleware;
      }

    });
    super(root, renderer);
  }

}

exports.default = App;
exports.ProviderPlugin = plugin;
exports.ProvidedHOC = hoc;
exports.Provider = Provider;
