// fusion-plugin-i18n-react/dist/browser.es5.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));
var i18n = require('fusion-plugin-i18n');
var i18n__default = _interopDefault(i18n);
var fusionReact = require('fusion-react');

function _inheritsLoose(subClass, superClass) { subClass.prototype.__proto__ = superClass && superClass.prototype; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var BundleSplitConsumer =
/*#__PURE__*/
function (_React$Component) {
  _inheritsLoose(BundleSplitConsumer, _React$Component);

  function BundleSplitConsumer(props, context) {
    var _this;

    _this = _React$Component.call(this, props, context) || this; // splitComponentLoaders comes from fusion-react-async/prepare-provider
    // ids comes from fusion-react-async/split
    // props.provides comes from fusion-react/plugin and references i18n()

    _this.i18n = props.provides.from(props.ctx);

    if (context.splitComponentLoaders) {
      context.splitComponentLoaders.push(function (ids) {
        return _this.i18n.load(ids);
      });
    }

    return _this;
  }

  var _proto = BundleSplitConsumer.prototype;

  _proto.getChildContext = function getChildContext() {
    return {
      i18n: this.i18n
    };
  };

  _proto.render = function render() {
    return React.Children.only(this.props.children);
  };

  return BundleSplitConsumer;
}(React.Component);

BundleSplitConsumer.contextTypes = {
  splitComponentLoaders: PropTypes.array
};
BundleSplitConsumer.childContextTypes = {
  i18n: PropTypes.object.isRequired
};
var I18n = fusionReact.ProviderPlugin.create('i18n', i18n__default, BundleSplitConsumer);

function _inheritsLoose$1(subClass, superClass) { subClass.prototype.__proto__ = superClass && superClass.prototype; subClass.__proto__ = superClass; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
/*
The `withTranslations` HOC takes an array of translation keys as an argument,
but does not use it at runtime.

However, these keys are captured by `babel-plugin-i18n` at compile-time by
Fusion's compiler and the compiler uses generate a map of all translations
in the app.

The translation map is then exposed by `fusion-plugin-i18n/chunk-translation-map.js`
*/

var withTranslations = function withTranslations()
/*translationKeys*/
{
  return function (OriginalComponent) {
    var WithTranslations =
    /*#__PURE__*/
    function (_Component) {
      _inheritsLoose$1(WithTranslations, _Component);

      function WithTranslations(props, context) {
        var _this;

        _this = _Component.call(this, props, context) || this;
        var i18n$$1 = context.i18n;
        _this.translateProp = i18n$$1 ? function (key, data) {
          return i18n$$1.translate(key, data);
        } : function (key) {
          return key;
        };
        return _this;
      }

      var _proto = WithTranslations.prototype;

      _proto.render = function render() {
        var finalProps = Object.assign({
          translate: this.translateProp
        }, this.props);
        return React__default.createElement(OriginalComponent, finalProps);
      };

      return WithTranslations;
    }(React.Component);

    WithTranslations.contextTypes = OriginalComponent.contextTypes = {
      i18n: PropTypes.object
    };
    WithTranslations.displayName = "withTranslations(" + (OriginalComponent.displayName || OriginalComponent.name) + ")";
    return WithTranslations;
  };
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
function Translate(props, context) {
  return React__default.createElement("span", null, context.i18n && context.i18n.translate(props.id, props.data) || props.id);
}

Translate.contextTypes = {
  i18n: PropTypes.object
};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
var withTranslationsTyped = withTranslations;

exports.I18nToken = i18n.I18nToken;
exports.I18nLoaderToken = i18n.I18nLoaderToken;
exports.HydrationStateToken = i18n.HydrationStateToken;
exports.createI18nLoader = i18n.createI18nLoader;
exports.default = I18n;
exports.withTranslations = withTranslationsTyped;
exports.Translate = Translate;
