// fusion-plugin-styletron-react/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var fusionCore = require('fusion-core');
var styletronReact = require('styletron-react');
var styletronEngineAtomic = require('styletron-engine-atomic');

/**
 * Provides styletron instance via old context API
 */

class LegacyStyletronProvider extends React.Component {
  getChildContext() {
    return {
      styletron: this.styletron
    };
  }

  constructor(props, context) {
    super(props, context);
    this.styletron = props.value;
  }

  render() {
    return React.Children.only(this.props.children);
  }

}

LegacyStyletronProvider.childContextTypes = {
  styletron: noop
};
function noop() {}

const injectDeclarationCompatMixin = Base => class extends Base {
  injectDeclaration({
    prop,
    val,
    media,
    pseudo
  }) {
    let style = {
      [prop]: val
    };

    if (pseudo) {
      style = {
        [pseudo]: style
      };
    }

    if (media) {
      style = {
        [media]: style
      };
    }

    return this.renderStyle(style);
  }

};

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */
const StyletronCompat = injectDeclarationCompatMixin(styletronEngineAtomic.Server);
const plugin = true && fusionCore.createPlugin({
  middleware: () => (ctx, next) => {
    if (ctx.element) {
      const engine = new StyletronCompat();
      ctx.element = React.createElement(styletronReact.Provider, {
        value: engine
      }, React.createElement(LegacyStyletronProvider, {
        value: engine
      }, ctx.element));
      return next().then(() => {
        const stylesForHead = engine.getStylesheetsHtml();
        ctx.template.head.push(fusionCore.dangerouslySetHTML(stylesForHead));
      });
    } else {
      return next();
    }
  }
});

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */
var index = plugin;

exports.styled = styletronReact.styled;
exports.withStyle = styletronReact.withStyle;
exports.withStyleDeep = styletronReact.withStyleDeep;
exports.withTransform = styletronReact.withTransform;
exports.withWrapper = styletronReact.withWrapper;
exports.default = index;
